added language filter

chreman opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR attempts to introduce a language filter as specified in the [CORE-API documentation](https://core.ac.uk/docs/), which to my understanding is so far missing from the advanced_search functionality.
## Description
<!--- Describe your changes in detail -->
I replicated as far as possible the structure and logic of existing search parameters, including a sanity check and the composition of the according elasticsearch query according to the CORE-API
## Related Issue
<!--- if this closes an issue make sure include e.g., "fix #4"
or similar - or if just relates to an issue make sure to mention
it like "#4" -->

## Example

```r
query <- data.frame("all_of_the_words" = "data mining",
                    "without_the_words" = "social science",
                    "year_from" = "2013",
                    "year_to" = "2014",
                    "language" = "English"
)

core_advanced_search(query)
```

chreman committed:
added language filter

sckott commented on R/advanced_search_helper_functions.R:
```diff
@@ -41,6 +41,16 @@ prepare_elasticsearch_term <- function(filter, filterName) {
+                      'zh', 'en', 'de', 'it', 'ja', 'ru', 'es')) {
+    return(TRUE)
+  } else {
+  return(FALSE)
```
indent this `return(FALSE)` inline with the `return(TRUE)` above

sckott commented on R/advanced_search_helper_functions.R:
```diff
@@ -41,6 +41,16 @@ prepare_elasticsearch_term <- function(filter, filterName) {
 }
 
+is_acceptable_language_filter <- function(language) {
+  if (language %in% c('Chinese', 'English', 'German', 'Italian', 'Japanese',
```
if the input `language` is always a single character string the comparison to the vector on the right with `%in%` already returns a boolean, so you could remove the if block

sckott reviewed (changes requested):
thanks @chreman 

a few comments for you

the big problem though is that your changes break the function. make sure to run tests for this function and add additional tests as needed to make sure your changes don't break the function AND give back what's expected (e.g. does querying by language in fact return results with that language)

chreman committed:
simplified language filter validity check

chreman committed:
added language filter test

chreman commented:
I simplified the language filter validity check as requested, and also added a test employing the language filter that ran with OK for all 12 tests - could you perhaps point me to where the function breaks?
In regards to whether it works as intended this may prove a bit more involving, since CORE does not provide language in the returned metadata we could compare against, and any automated test would involve language parsing. But I will look into possibilities.

sckott commented:
thanks @chreman for the changes

can you add an entry for `language` here https://github.com/ropensci/rcoreoa/blob/master/R/core_advanced_search.R#L33 following format of the others

sckott commented:
> in regards to whether it works as intended this may prove a bit more involving

okay, we can skip that for now

sckott commented:
We i think just need to add 

```r
languageFilter <- prepare_elasticsearch_term(query["language"], "language.name")
```

here https://github.com/ropensci/rcoreoa/blob/master/R/advanced_search_helper_functions.R#L187 

and it should work. witthout that it was leaving a trailing `AND` in the query which was causing the query to crash on their end

chreman committed:
languageFilter now elasticsearch_term

chreman committed:
added to docstring

chreman committed:
cleaned up accepted languages; added new elasticsearch_term builder

chreman committed:
fixed query when no language asked

chreman commented:
Thank you for the hints. I added a short docstring; removed incompatible languages from the acceptable_filter; and introduced a new elasticsearch_term preparation function, as the existing one was inappropriate - the query fails when the language is enclosed in parentheses [api-doc](https://core.ac.uk/docs/).

I also did a few simple queries for `baguette` in the data range from 1900 to 2018 in `English, German, French, and Spanish` and indeed the function returned mainly documents of that language - although in a much lower quantity.

sckott commented:
thanks for this @chreman  - on vacation now, will respond mid next week

sckott commented:
thanks @chreman looks good now
